File: .validate/scripts/convert-value-to-children.ts
#!/usr/bin/env bun
/** One-off: convert @ui.Text(p, value) to @ui.Text(p) { value } */
import { readdirSync, readFileSync, writeFileSync, statSync } from "fs";
import { join, relative } from "path";

const COMPONENTS = [
  ["ui", "Text"],
  ["ui", "Inline"],
  ["ui", "Title"],
  ["ui", "FormDescription"],
  ["ui", "FormMessage"],
  ["cmp", "CardTitle"],
  ["cmp", "CardDescription"],
  ["cmp", "SheetTitle"],
  ["cmp", "SheetDescription"],
] as const;

function depthDelta(ch: string): number {
  if ("({".includes(ch)) return 1;
  if (")}".includes(ch)) return -1;
  return 0;
}

function convertContent(text: string): { text: string; changes: number } {
  let changes = 0;
  for (const [prefix, name] of COMPONENTS) {
    const needle = `@${prefix}.${name}(`;
    let pos = 0;
    while (true) {
      const start = text.indexOf(needle, pos);
      if (start === -1) break;
      let i = start + needle.length;
      let depth = 1;
      let inString = false;
      let stringQuote = "";
      while (i < text.length && depth > 0) {
        const ch = text[i];
        if (inString) {
          if (ch === stringQuote && text[i - 1] !== "\\") inString = false;
        } else if ('"`\''.includes(ch)) {
          inString = true;
          stringQuote = ch;
        } else {
          depth += depthDelta(ch);
        }
        i++;
      }
      if (depth !== 0) {
        pos = start + needle.length;
        continue;
      }
      const inner = text.slice(start + needle.length, i - 1);
      let splitAt = -1;
      depth = 0;
      inString = false;
      stringQuote = "";
      for (let j = 0; j < inner.length; j++) {
        const ch = inner[j];
        if (inString) {
          if (ch === stringQuote && (j === 0 || inner[j - 1] !== "\\")) inString = false;
        } else if ('"`\''.includes(ch)) {
          inString = true;
          stringQuote = ch;
        } else {
          if ("({".includes(ch)) depth++;
          else if (")}".includes(ch)) depth--;
          else if (ch === "," && depth === 0) splitAt = j;
        }
      }
      if (splitAt === -1) {
        pos = i;
        continue;
      }
      const args = inner.slice(0, splitAt).trim();
      const value = inner.slice(splitAt + 1).trim();
      const replacement = `@${prefix}.${name}(${args}) { ${value} }`;
      text = text.slice(0, start) + replacement + text.slice(i);
      changes++;
      pos = start + replacement.length;
    }
  }
  return { text, changes };
}

function walk(dir: string, acc: string[] = []): string[] {
  for (const ent of readdirSync(dir)) {
    const p = join(dir, ent);
    if (statSync(p).isDirectory()) acc = walk(p, acc);
    else acc.push(p);
  }
  return acc;
}

const root = join(import.meta.dir, "../..");
const roots = ["ui", "components", join("examples", "templ")];
let total = 0;
for (const r of roots) {
  const base = join(root, r);
  for (const path of walk(base)) {
    if (!path.endsWith(".templ") && !path.endsWith(".spec.md")) continue;
    if (path.includes("_templ.go")) continue;
    const original = readFileSync(path, "utf8");
    const { text, changes } = convertContent(original);
    if (changes > 0) {
      writeFileSync(path, text.replace(/\r\n/g, "\n"), "utf8");
      console.log(`${relative(root, path)}: ${changes}`);
      total += changes;
    }
  }
}
console.log(`done: ${total}`);

File: .validate/scripts/convert-value-to-children.py
#!/usr/bin/env python3
"""One-off: convert @ui.Text(p, value) to @ui.Text(p) { value } for parity sweep."""
from __future__ import annotations

import re
import sys
from pathlib import Path

COMPONENTS = [
    ("ui", "Text"),
    ("ui", "Inline"),
    ("ui", "Title"),
    ("ui", "FormDescription"),
    ("ui", "FormMessage"),
    ("cmp", "CardTitle"),
    ("cmp", "CardDescription"),
    ("cmp", "SheetTitle"),
    ("cmp", "SheetDescription"),
]


def depth_delta(ch: str) -> int:
    if ch in "({":
        return 1
    if ch in ")}":
        return -1
    return 0


def convert_content(text: str) -> tuple[str, int]:
    changes = 0
    for prefix, name in COMPONENTS:
        needle = f"@{prefix}.{name}("
        pos = 0
        while True:
            start = text.find(needle, pos)
            if start == -1:
                break
            i = start + len(needle)
            depth = 1
            in_string = False
            string_quote = ""
            while i < len(text) and depth > 0:
                ch = text[i]
                if in_string:
                    if ch == string_quote and text[i - 1] != "\\":
                        in_string = False
                elif ch in "\"'`":
                    in_string = True
                    string_quote = ch
                else:
                    depth += depth_delta(ch)
                i += 1
            if depth != 0:
                pos = start + len(needle)
                continue
            inner = text[start + len(needle) : i - 1]
            split_at = -1
            depth = 0
            in_string = False
            string_quote = ""
            for j, ch in enumerate(inner):
                if in_string:
                    if ch == string_quote and (j == 0 or inner[j - 1] != "\\"):
                        in_string = False
                elif ch in "\"'`":
                    in_string = True
                    string_quote = ch
                else:
                    if ch in "({":
                        depth += 1
                    elif ch in ")}":
                        depth -= 1
                    elif ch == "," and depth == 0:
                        split_at = j
            if split_at == -1:
                pos = i
                continue
            args = inner[:split_at].strip()
            value = inner[split_at + 1 :].strip()
            replacement = f"@{prefix}.{name}({args}) {{ {value} }}"
            text = text[:start] + replacement + text[i:]
            changes += 1
            pos = start + len(replacement)
    return text, changes


def main() -> int:
    root = Path(__file__).resolve().parents[2]
    globs = [
        "ui/**/*.templ",
        "ui/**/*.spec.md",
        "components/**/*.templ",
        "components/**/*.spec.md",
        "examples/templ/**/*.templ",
    ]
    total = 0
    for pattern in globs:
        for path in root.glob(pattern):
            if path.name.endswith("_templ.go"):
                continue
            original = path.read_text(encoding="utf-8")
            converted, n = convert_content(original)
            if n:
                path.write_text(converted, encoding="utf-8", newline="\n")
                print(f"{path.relative_to(root)}: {n} conversion(s)")
                total += n
    print(f"done: {total} conversion(s)")
    return 0


if __name__ == "__main__":
    sys.exit(main())
